### PR TITLE
[GEP-36] Self-Hosted Shoot Exposure extend Shoot API

### PR DESCRIPTION
**Pull request by robinschneider**
**How to categorize this PR?**

/area ipcei networking
/kind enhancement

**What this PR does / why we need it**:
This PR extends the Shoot API for self-hosted shoots with managed infrastructure for exposure extensions.

**Which issue(s) this PR fixes**:
Part of https://github.com/gardener/gardener/issues/13602

**Special notes for your reviewer**:
/cc @rfranzke @ScheererJ

**Release note**:

```other operator
NONE
```

### CODE REVIEW

File: pkg/apis/core/v1beta1/types_shoot.go [hunk 1]
```diff
@@ -1809,8 +1809,38 @@ type WorkerControlPlane struct {
 	// If it is not specified, then there won't be any backups taken.
 	// +optional
 	Backup *Backup `json:"backup,omitempty" protobuf:"bytes,1,opt,name=backup"`
+	// Exposure holds the exposure configuration for the shoot (either `extension` or `dns` or null) [GEP36].
```
Comment (ScheererJ) [inline pkg/apis/core/v1beta1/types_shoot.go]: What do you mean with `null`? In `go` the zero value for pointers is called `nil`.

Not sure if we really want to annotate the code with `[GEP36]`. We usually do not do this.

File: pkg/apis/core/v1beta1/types_shoot.go [hunk 2]
```diff
@@ -1809,8 +1809,38 @@ type WorkerControlPlane struct {
...
+	// +optional
+	Exposure *Exposure `json:"exposure,omitempty" protobuf:"bytes,2,opt,name=exposure"`
+}
+
+// Exposure holds the exposure configuration for the shoot (either `extension` or `dns` or null) [GEP36].
+type Exposure struct {
```
Comment (ScheererJ) [inline pkg/apis/core/v1beta1/types_shoot.go]: While I understand that this comment is copied from above, I have trouble understand what `null` means here even if I read it as `nil`. Does it mean that the `Exposure` field in the `WorkerControlPlane` is `nil` or should it mean an empty `Exposure{}`?

File: example/90-shoot.yaml [hunk 3]
```diff
@@ -161,6 +161,11 @@ spec:
     #       kind: Secret
     #       name: backup-secret
     #       namespace: kube-system
+    #  exposure: # either `extension` or `dns` or null
```
Comment (ScheererJ) [inline example/90-shoot.yaml]: What does `null` mean here?

File: pkg/apis/core/validation/shoot.go [hunk 4]
```diff
@@ -2688,6 +2666,40 @@ func ValidateWorkers(workers []core.Worker, fldPath *field.Path) field.ErrorList
 	return allErrs
 }
 
+func ValidateWorkerControlPlane(plane *core.WorkerControlPlane, shootNamespace, shootProviderType string, fldPath *field.Path) field.ErrorList {
```
Comment (ScheererJ) [inline pkg/apis/core/validation/shoot.go]:
```suggestion
func ValidateWorkerControlPlane(controlPlane *core.WorkerControlPlane, shootNamespace, shootProviderType string, fldPath *field.Path) field.ErrorList {
```

File: pkg/apis/core/validation/shoot.go [hunk 5]
```diff
@@ -2688,6 +2666,40 @@ func ValidateWorkers(workers []core.Worker, fldPath *field.Path) field.ErrorList
...
+				allErrs = append(allErrs, field.Forbidden(fldPath.Child("backup", "credentialsRef", "namespace"), "must reference credentials in the same namespace as the Shoot"))
+			}
+
+			// TODO(vpnachev): Allow WorkloadIdentities once the support in the controllers and components is fully implemented.
+			if backup.CredentialsRef.APIVersion == securityv1alpha1.SchemeGroupVersion.String() && backup.CredentialsRef.Kind == "WorkloadIdentity" {
+				allErrs = append(allErrs, field.Forbidden(fldPath.Child("backup", "credentialsRef"), "support for WorkloadIdentity as backup credentials is not yet implemented"))
+			}
+		}
+	}
+	if plane.Exposure != nil {
```
Comment (ScheererJ) [inline pkg/apis/core/validation/shoot.go]:
```suggestion
	}

	if plane.Exposure != nil {
```

File: pkg/apis/core/validation/shoot.go [hunk 6]
```diff
@@ -2688,6 +2666,40 @@ func ValidateWorkers(workers []core.Worker, fldPath *field.Path) field.ErrorList
...
+		if plane.Exposure.DNS != nil && plane.Exposure.Extension != nil {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("exposure"), plane.Exposure, "cannot have dns and extension set at the same time"))
+		}
+	}
+	return allErrs
```
Comment (ScheererJ) [inline pkg/apis/core/validation/shoot.go:2712]:
```suggestion
	}

	return allErrs
```

File: pkg/apis/core/validation/shoot.go [hunk 7]
```diff
@@ -2688,6 +2666,40 @@ func ValidateWorkers(workers []core.Worker, fldPath *field.Path) field.ErrorList
...
+			// TODO(vpnachev): Allow WorkloadIdentities once the support in the controllers and components is fully implemented.
+			if backup.CredentialsRef.APIVersion == securityv1alpha1.SchemeGroupVersion.String() && backup.CredentialsRef.Kind == "WorkloadIdentity" {
+				allErrs = append(allErrs, field.Forbidden(fldPath.Child("backup", "credentialsRef"), "support for WorkloadIdentity as backup credentials is not yet implemented"))
+			}
+		}
+	}
+	if plane.Exposure != nil {
+		if plane.Exposure.DNS != nil && plane.Exposure.Extension != nil {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("exposure"), plane.Exposure, "cannot have dns and extension set at the same time"))
+		}
```
Comment (ScheererJ) [inline pkg/apis/core/validation/shoot.go:2709]: Please add validation for the `type` field in `plane.Exposure.Extension` similar to https://github.com/gardener/gardener/blob/574e7a7a23d835d08b0029d2dacbaaf338aee82a/pkg/apis/core/validation/shoot.go#L2066-L2070, but allowing empty `type`.

File: pkg/apis/core/validation/shoot_test.go [hunk 8]
```diff
@@ -1864,6 +1864,54 @@ var _ = Describe("Shoot Validation Tests", func() {
...
+						Exposure: &core.Exposure{
+							Extension: &core.ExtensionExposure{
+								Type:           ptr.To("local"),
+								ProviderConfig: &runtime.RawExtension{Raw: []byte(`{"apiVersion": "v1", "kind": "Configuration"}`)},
+							},
+						},
+					}
+
+					Expect(ValidateShoot(shoot)).To(BeEmpty())
+				})
```
Comment (ScheererJ) [inline pkg/apis/core/validation/shoot_test.go:1973]: Please also add a test with an empty `Exposure{}`.

**Review by ScheererJ — COMMENTED** [review]
Thanks for starting the work on GEP-36 to exposure self-hosted shoot clusters on managed infrastructures.

File: example/90-shoot.yaml [hunk 9]
```diff
@@ -161,6 +161,11 @@ spec:
     #       kind: Secret
     #       name: backup-secret
     #       namespace: kube-system
+    #  exposure: # either `extension` or `dns` or omitted/empty
+    #    extension:
+    #      type: local # defaults to `.spec.provider.type`, but could also be different
+    #      providerConfig: {} # *runtime.RawExtension
+    #    dns: {}
```
Comment (timebertt) [inline example/90-shoot.yaml]: I believe 1 space is missing for consistent indentation.
```suggestion
    #   exposure: # either `extension` or `dns` or omitted/empty
    #     extension:
    #       type: local # defaults to `.spec.provider.type`, but could also be different
    #       providerConfig: {} # *runtime.RawExtension
    #     dns: {}
```

**Review by timebertt — COMMENTED** [review]
Nice PR, well done :)

**Review by timebertt — COMMENTED** [review]
/lgtm

**Review by rfranzke — COMMENTED** [review]
/approve
